### DISCUSSION TITLE
ULP on ESP32 NodeMCU power consumption

### DISCUSSION THREAD

**Discussion by belveder79** [arendst/Tasmota#19276] [Q&A]
I'm playing around with the ULP feature on a NodeMCU, actually [this](https://www.amazon.de/diymore-Entwicklungsplatine-NodeMCU-Module-Bluetooth/dp/B09Z6RLS7R/ref=sr_1_3?keywords=esp32&qid=1691412675&sr=8-3&th=1) one.

First, the ULP module seems to do what I want, which means that it turns off if a GPIO is down, as if it pulled up, it loads (making a simple keybox open/close mechanism with an opener switch). 
I'm driving it with an 18650 4.1V LiPo battery over a USB shield, and I was actually expecting it to last kind-of forever. However, it turns out that it still drains pretty quickly, like 0.1V per day. Unfortunately I was not able to measure real power consumption with a multimeter yet, but I'll try to do so soon. 

Overall I wonder about two things even before doing that:

- the ESP32 is obviously the main source of power drain if it is on with WIFI, BT, whatever. However, if I use the regular micro-usb connector does the USB-Serial peripherals still draw significant current even if the ESP32 is in ULP.sleep()? I believe this is indeed the case...
- the question is then, is there a way AT ALL to drive a NodeMCU with ULP with kind-of zero power draw? My understanding is that I can obviously use the NodeMCU pins straight to provide power to the ESP32, but doesn't this again also power the USB-serial circuit anyway? This would mean to me that the ULP mode only really makes sense if I have a custom (aka minimalistic) logic assembled around a plain ESP32 on a custom board to make real use of the ULP mode, right?

Any thoughts? Thx...

**Comment by joba-1** [comment]
Ok, some thoughts:

Basically all your assumptions are correct. The onboard LDO,, serial chip etc. all draw way too much power for a battery powered device. Just to give some ballpark numbers: Wifi/Bt ~100mA, ESP32 w/o RF: 10mA, ULP: 1mA, add for the board at least 2mA.
Also USB shields draw additional power. Like that you hardly get run times of more than a few days.
What you need is low µA, not mA.

I wanted to do battery based ESP32 ULP to read and collect sensor data via I2C and send in bulk every now and then at one point and gave up this plan.

It could get much better with a specialized board (trigboard comes to mind), but they are rare and not cheap). Also there are some remotes using ESP Now for sending, so it is doable. Maybe that works for you, I didn't test it.
Instead I used low power versions of STM32 chips (STM32L0...) and a LoRa chip (RFW95) for sending/receiving data. Only the Gateway is an ESP, also with an RFW95. 

For the STM32/RFW95 boards I used, only the sending power and duration is relevant for battery life. In relation to that all else is peanuts. Since sending power is quite high (you can't beat physics), better send data as bytes, not as text (like json or so) to make the transmission as short as possible.
Like that I achieve run times of over 2 years and counting with 2 AAA cells (~50ms sending every 5min). My main enemy is corrosion, not battery life :)

For your use case I don't think you need ULP. The ESP (or STM32) can deep sleep completely until a pin level change from a contact wakes it up via interrupt. It would basically run forever (ok, maybe 10 years, similar to batteries when not used).

Second to last thought: expect problems debugging unexpected power consumptions. Normal multimeters can't measure them. I basically was lucky it worked as planned but a device like the µCurrent might become necessary.

Last thought: rechargeable batteries loose power much faster than "normal" batteries even while the circuit is off. I would only use them in conjunction with a small solar panel and charging circuit.

**Comment by Jason2866** [comment]
It is possible to run an ESP32 for month waking up a short time and transmit data via wifi.
ULP needs a few uA! not mA.
But this is NOT possible with a standard development board. The board needs to be designed for battery use.
I have a Weather station with a ePaper Display waking up every 10 minutes. The LiPo 18650 holds 6 months until charging is needed.

**Comment by belveder79** [comment]
Thx a lot for all the insights - good to know that my previous studies in electronics weren't totally wasted :-)

I've ordered a few plain ESP32 boards without supplemental stuff and will check how those behave, but yes, it makes total sense to "design for battery" and leave out all the unnecessary stuff around... I'm still somewhat flashed by the way the Tuya devices work as discussed in https://github.com/arendst/Tasmota/discussions/18935. They just turn off the ESP entirely by cutting power to it from an MCU. Overall I think it is the cleverest way to reduce power consumption for smoke detectors or window/door sensors, at the expense of having another dedicated microcontroller, however...

will post updates on this once the boards arrive...

**Reply by sfromis** [reply]
Unless you want to design your own, I'd suggest getting a board already having the hardware for battery support, without this meaning removing other features. Examples are boards from Unexpected Maker or LilyGo.
